### DISCUSSION TITLE
testing: new object in array after callback

### DISCUSSION THREAD

**Discussion by DanielGennat** [neuefische/questions#236] [Web] [answered]
Hi, I wrote two attempts to test my handleAdd function and I don't understand why I get this type error both times and how to fix it:

The function I want to test:
```
//EditTimerChainForm.js
export function handleAdd(newTimerChain, setNewTimerChain) {
  const addId = newTimerChain.length + 1;
  setNewTimerChain([...newTimerChain, { id: addId, minutes: 5, seconds: 0 }]);
}
```
The tests I wrote:
```
//handleAdd.test.js
import handleAdd from '../components/EditFormComponents/EditTimerChainForm';

describe('newTimerChain', () => {
  test('has new timer with 5 minutes', () => {
    let newTimerChain = [{ id: 1, minutes: 1, seconds: 10 }];
    const setNewTimerChain = jest.fn();
    const addId = newTimerChain.length + 1;
    handleAdd(newTimerChain, setNewTimerChain);
    const newTimerChainAfterHandleAdd = [
      { id: 1, minutes: 1, seconds: 10 },
      { id: addId, minutes: 5, seconds: 0 },
    ];
    expect(newTimerChain).toMatchObject(newTimerChainAfterHandleAdd);
  });
});

describe('newTimerChain', () => {
  test('has new timer with 5 minutes', () => {
    let newTimerChain = [{ id: 1, minutes: 1, seconds: 10 }];
    const setNewTimerChain = jest.fn();
    const addId = newTimerChain.length + 1;
    const newTimerChainAfterHandleAdd = [
      { id: 1, minutes: 1, seconds: 10 },
      { id: addId, minutes: 5, seconds: 0 },
    ];
    handleAdd(newTimerChain, setNewTimerChain);
    expect(setNewTimerChain).toHaveBeenLastCalledWith(
      newTimerChainAfterHandleAdd
    );
  });
});
```
The error Message I got both times:
<img width="445" alt="Bildschirmfoto 2022-10-10 um 11 30 36" src="https://user-images.githubusercontent.com/109812125/194836535-b3301c55-3c3f-4f1c-8fb9-bbcacb892c8d.png">

My [PR](https://github.com/DanielGennat/capstone-project/pull/13)

**Comment by NikRoe** [comment, marked as answer]
Hi Daniel,

you have a named export and your import needs to look like this:
```js
import { handleAdd } from '../components/EditFormComponents/EditTimerChainForm';
```

With your first test you would be testing useState itself and that is not necessary. 

Your second test looks a lot better. However, you don't need to extract parts of the logic.
`const addId = newTimerChain.length + 1;` is not needed here. Your `handleAdd()` takes care of this.

The last missing part is that you can work with static data from the start if you want to compare them. This will do just fine:
```js
const newTimerChainAfterHandleAdd = [
      { id: 1, minutes: 1, seconds: 10 },
      { id: 2, minutes: 5, seconds: 0 },
    ];
```

**Reply by DanielGennat** [reply]
So just the named import (and some redundand code).
Thanks! :)
